opnode/l1/downloader.go

```diff
@@ -0,0 +1,271 @@
+package l1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+const (
+	fetchBlockTimeout   = 10 * time.Second
+	fetchReceiptTimeout = 10 * time.Second
+	maxReceiptRetry     = 5
+
+	// 500 at 100 KB each would be 50 MB of memory for the L1 block inputs cache
+	cacheSize = 500
+
+	// Amount of receipt tasks to buffer before applying back-pressure (blocking new block-requests)
+	receiptQueueSize = 100
+)
+
+var (
+	DownloadEvictedErr = errors.New("evicted")
+	DownloadClosedErr  = errors.New("closed")
+)
+
+type downloadTask struct {
+	// Incremented when Finish is called, to check whether the task already completed (with or without error).
+	// May increment >1 when other sub-tasks fail.
+	//
+	// First field, aligned atomic changes.
+	finished uint32
+
+	// Count already downloaded receipts to track status
+	//
+	// Aligned after above field, atomic changes.
+	downloadedReceipts uint32
+
+	block *types.Block
+	// receipts slice is allocated in advance, one slot for each transaction, nil until downloaded
+	receipts []*types.Receipt
+
+	// stop fetching if this context is dead
+	ctx context.Context
+	// to free up resources (e.g. ongoing receipt tasks) if a downloadTask finishes prematurely
+	cancel context.CancelFunc
+
+	// feed to subscribe the requests to (de-duplicate work)
+	feed event.Feed
```

Comment:
If I'm reading the code correctly, this is currently just used to signal to the Fetch function that it has succeeded or failed. If so, it feels a little overkill.

Reply:
Do you suggest an alternative to signal completion?

Reply:
It's not, we can't use a regular channel since there may be multiple listeners waiting for the job (duplicates, as described in the comment).

Reply:
It makes sense to use the feed in this case, but I don't love a design that requires a queue to de-duplicate in the first place